File: src/main/java/sage/web/ListController.java
package sage.web;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.util.Assert;
import org.springframework.web.bind.annotation.*;
import sage.domain.service.ListService;
import sage.domain.service.RelationService;
import sage.domain.service.TagService;
import sage.domain.service.UserService;
import sage.entity.Tag;
import sage.transfer.*;
import sage.util.Colls;
import sage.web.auth.Auth;
import sage.web.context.Json;

@RestController
@RequestMapping("/list")
public class ListController {
  @Autowired
  private ListService listService;
  @Autowired
  private RelationService relationService;
  @Autowired
  private UserService userService;
  @Autowired
  private TagService tagService;
  
  @RequestMapping(value = "/resource/{id}", method = RequestMethod.GET)
  public ResourceList getResourceList(@PathVariable Long id) {
    Auth.checkCuid();
    return listService.getResourceList(id);
  }
  
  @RequestMapping(value = "/resource/{id}", method = RequestMethod.POST)
  public Boolean updateResourceList(@PathVariable Long id, @RequestParam String list) {
    Long uid = Auth.checkCuid();
    
    ResourceList rc = Json.object(list, ResourceList.class);
    Assert.isTrue(rc.getOwnerId().equals(uid));
    Assert.isTrue(rc.getId().equals(id));
    listService.updateResourceList(rc, uid);
    return true;
  }

  @RequestMapping(value = "/resource/add", method = RequestMethod.POST)
  public Long addResourceList(@RequestParam String list) {
    Long uid = Auth.checkCuid();
    
    ResourceList rc = Json.object(list, ResourceList.class);
    return listService.addResourceList(rc, uid);
  }
  
  @RequestMapping(value = "/follow/{id}", method = RequestMethod.GET)
  public FollowList getFollowList(@PathVariable Long id) {
    Auth.checkCuid();
    return listService.getFollowList(id);
  }
  
  @RequestMapping(value = "/follow/{id}", method = RequestMethod.POST)
  public Boolean updateFollowList(@PathVariable Long id, @RequestParam String listLite) {
    Long uid = Auth.checkCuid();
    
    FollowListLite fcLite = Json.object(listLite, FollowListLite.class);
    Assert.isTrue(fcLite.getOwnerId().equals(uid));
    Assert.isTrue(fcLite.getId().equals(id));
    listService.updateFollowList(fcLite, uid);
    return true;
  }

  @RequestMapping(value = "/follow/add", method = RequestMethod.POST)
  public Long addFollowList(@RequestParam String listLite) {
    Long uid = Auth.checkCuid();

    FollowListLite fcLite = Json.object(listLite, FollowListLite.class);
    fcLite.setOwnerId(uid);
    return listService.addFollowList(fcLite, uid);
  }

  @RequestMapping(value = "/follow/expose-all", method = RequestMethod.POST)
  public Long exposeAllOfFollow() {
    Long cuid = Auth.checkCuid();

    List<FollowInfoLite> follows = Colls.map(relationService.followings(cuid), f ->
        new FollowInfoLite(f.getTarget().getId(), Colls.map(f.getTags(), Tag::getId)));
    FollowListLite list = new FollowListLite(null, cuid, "所有关注", follows);
    return listService.addFollowList(list, cuid);
  }
}

File: src/main/kotlin/sage/web/api/ListController.kt
package sage.web.api

import org.springframework.web.bind.annotation.*
import sage.domain.commons.DomainException
import sage.transfer.*
import sage.web.auth.Auth
import sage.web.context.BaseController
import sage.util.Json

@RestController
@RequestMapping("/list")
class ListController : BaseController() {

  @GetMapping("/resource/{id}")
  fun getResourceList(@PathVariable id: Long): ResourceList {
    Auth.checkUid()
    return listService.getResourceList(id)
  }

  @PostMapping("/resource/{id}")
  fun updateResourceList(@PathVariable id: Long, @RequestParam list: String): Boolean {
    val uid = Auth.checkUid()

    val rc = Json.`object`(list, ResourceList::class.java)
    checkBeforeUpdate(id, rc)
    listService.updateResourceList(rc, uid)
    return true
  }

  @PostMapping("/resource/add")
  fun addResourceList(@RequestParam list: String): Long {
    val uid = Auth.checkUid()

    val rc = Json.`object`(list, ResourceList::class.java)
    return listService.addResourceList(rc, uid)
  }

  @GetMapping("/follow/{id}")
  fun getFollowList(@PathVariable id: Long): FollowList {
    Auth.checkUid()
    return listService.getFollowList(id)
  }

  @PostMapping("/follow/{id}")
  fun updateFollowList(@PathVariable id: Long, @RequestParam listLite: String): Boolean {
    val uid = Auth.checkUid()

    val fcLite = Json.`object`(listLite, FollowListLite::class.java)
    checkBeforeUpdate(id, fcLite)
    listService.updateFollowList(fcLite, uid)
    return true
  }

  @PostMapping("/follow/add")
  fun addFollowList(@RequestParam listLite: String): Long {
    val uid = Auth.checkUid()

    val fcLite = Json.`object`(listLite, FollowListLite::class.java)
    fcLite.ownerId = uid
    return listService.addFollowList(fcLite, uid)
  }

  @PostMapping("/follow/expose-all")
  fun exposeAllOfFollow(): Long {
    val uid = Auth.checkUid()

    val follows = relationService.followings(uid).map { f -> FollowInfoLite(f.target.id, f.tags.map { it.id }) }

    val existingIdOrNull = listService.followListsOfOwner(uid).find { it.name == "所有关注" }?.id
    val list = FollowListLite(existingIdOrNull, uid, "所有关注", follows)

    return if (existingIdOrNull == null) {
      listService.addFollowList(list, uid)
    } else {
      listService.updateFollowList(list, uid)
    }
  }

  private fun checkBeforeUpdate(id: Long, obj: AList) {
    if (obj.id != id) {
      throw DomainException("id不匹配：链接中id=${id}，对象中id=${obj.id}")
    }
    val uid = Auth.uid()
    if (obj.ownerId != uid) {
      throw DomainException("User[$uid] is not allowed to edit ${obj.javaClass}[$id]")
    }
  }
}
